### PR TITLE
Add com.ultimaker.cura

### PR DESCRIPTION
**Pull request by Tim453**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link: https://github.com/Ultimaker/Cura/issues/6027#issuecomment-840809524**
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(All patches are related to the flatpak build-process)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: com.ultimaker.cura.yml [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+  - --share=network
+  # For accessing usb printers
+  - --device=all
+  # To access user files
+  - --filesystem=home
+command: cura
+cleanup:
+  - /include
+  - /lib/cmake
+  - /lib/debug
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:21]:
```suggestion
```

Debug symbols are split out in a debug package. Don't remove them

File: com.ultimaker.cura.yml [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+      - mv /app/share/mime/packages/cura.xml /app/share/mime/packages/com.ultimaker.cura.xml
+    sources:
+      - type: git
+        url: https://github.com/ultimaker/Cura
+        commit: c328f3619dae3ad37fb1decccf1013afa18b410c
+        tag: 4.13.0
+
+
+  - name: cura-binary-data
+    buildsystem: cmake
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:306]:
```suggestion
    buildsystem: cmake-ninja
```

unless it fails, it's faster

File: com.ultimaker.cura.yml [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+      - type: git
+        url: https://github.com/ultimaker/fdm_materials
+        commit: 304287c3a69d79a14d33c1fe897bf653a2fc4f15
+        tag: 4.13.0
+
+  - name: cura
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DCMAKE_INSTALL_PREFIX=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:286]:
```suggestion
```

These are automatically set

File: com.ultimaker.cura.yml [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    sources:
+      - type: git
+        url: https://github.com/ultimaker/fdm_materials
+        commit: 304287c3a69d79a14d33c1fe897bf653a2fc4f15
+        tag: 4.13.0
+
+  - name: cura
+    buildsystem: cmake
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:283]:
```suggestion
    buildsystem: cmake-ninja
```

unless it fails, it's faster

File: com.ultimaker.cura.yml [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+      - type: git
+        url: https://github.com/ultimaker/Uranium
+        commit: b1d9219a7272140a2021e7339901897c7df927d5
+        tag: 4.13.0
+
+  - name: fdm_materials
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DCMAKE_INSTALL_PREFIX=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:275]:
```suggestion
```

File: com.ultimaker.cura.yml [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    sources:
+      - type: git
+        url: https://github.com/ultimaker/Uranium
+        commit: b1d9219a7272140a2021e7339901897c7df927d5
+        tag: 4.13.0
+
+  - name: fdm_materials
+    buildsystem: cmake
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:272]:
```suggestion
    buildsystem: cmake-ninja
```

unless it fails, it's faster

File: com.ultimaker.cura.yml [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        url: https://github.com/ultimaker/CuraEngine
+        commit: 0e79559492a9e3e631c28624b73a888b3297f174
+        tag: 4.13.0
+
+
+  - name: uranium
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DCMAKE_INSTALL_PREFIX=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:264]:
```suggestion
    buildsystem: cmake-ninja
```

File: com.ultimaker.cura.yml [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+        url: https://github.com/Ultimaker/libCharon
+        commit: 8e255aaad10b09c9cfc6e795f33968fd642e26c9
+        tag: 4.13.0
+
+  - name: cura-engine
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=Release
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DCMAKE_INSTALL_PREFIX=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:248]:
```suggestion
    buildsystem: cmake-ninja
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

File: com.ultimaker.cura.yml [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+        commit: 68a8868e0b30883b461c0dde77f6e85ce0a4d2ef
+        tag: 4.13.0
+      - type: patch
+        path: patch/libsavitar-python-install.patch
+
+  - name: libCharon
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DCMAKE_INSTALL_PREFIX=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:235]:
```suggestion
    buildsystem: cmake-ninja
    config-opts:
```

File: com.ultimaker.cura.yml [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+      - -DPYTHON_EXECUTABLE=/app/bin/python3
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:236]: are you sure it's the right path?

File: com.ultimaker.cura.yml [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+        tag: 4.12.1
+      - type: patch
+        path: patch/pynest2d-python-install.patch
+
+  - name: savitar
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=Release
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DCMAKE_INSTALL_PREFIX=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:220]:
```suggestion
    buildsystem: cmake-ninja
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

File: com.ultimaker.cura.yml [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+        tag: 4.13.0 
+      - type: patch
+        path: patch/libarcus-python-binindg-install.patch
+
+  - name: pynest2d
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=Release
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DCMAKE_INSTALL_PREFIX=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:204]:
```suggestion
    buildsystem: cmake-ninja
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

File: com.ultimaker.cura.yml [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+      - type: git
+        url: https://github.com/nothings/stb.git
+        commit: d5d052c806eee2ca1f858cb58b2f062d9fa25b90
+
+  - name: arcus
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=Release
+      - -DCMAKE_INSTALL_PREFIX=/app
+      - -DCMAKE_PREFIX_PATH=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:186]:
```suggestion
    buildsystem: cmake-ninja
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

File: com.ultimaker.cura.yml [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+          - mv FindClipper.cmake cmake_modules/
+
+  - name: openctm
+    buildsystem: simple
+    build-commands:
+      - make -f Makefile.linux openctm
+      - cp lib/libopenctm.so /app/lib/libopenctm.so
+      - cp lib/openctm.h     /app/include/openctm.h
+      - cp lib/openctmpp.h   /app/include/openctmpp.h
+    sources:
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:159]:
```suggestion
    cleanup:
      - /include
    sources:
```

File: com.ultimaker.cura.yml [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+        url: https://raw.githubusercontent.com/Ultimaker/cura-build-environment/4.12/projects/clipper_cmakelists_patch.txt
+        sha256: 54c430466bcbe745fc85b757c02a63868d26022891cc1b98fb9862e9e20c79fb
+        dest-filename: CMakeLists.txt
+
+  - name: libnest2d
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=Release
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DCMAKE_INSTALL_PREFIX=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:138]:
```suggestion
    buildsystem: cmake-ninja
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

File: com.ultimaker.cura.yml [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+      - type: archive
+        url: http://sourceforge.net/projects/boost/files/boost/1.67.0/boost_1_67_0.tar.bz2
+        sha256: 2684c972994ee57fc5632e03bf044746f6eb45d4920c343937a465fd67a5adba
+
+  - name: clipper
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=Release
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DCMAKE_INSTALL_PREFIX=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:121]:
```suggestion
    buildsystem: cmake-ninja
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

File: com.ultimaker.cura.yml [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+        url: https://github.com/stevengj/nlopt.git
+        commit: 41967f1981d82d8495c0b27151a126bc35ae0d96
+        tag: v2.6.2
+
+
+  - name: boostheaders
+    buildsystem: simple
+    build-commands:
+      - cp -r boost /app/include/boost
+    sources:
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:111]:
```suggestion
    cleanup:
      - /include
    sources:
```

File: com.ultimaker.cura.yml [hunk 18]
```diff
@@ -0,0 +1,313 @@
...
+      - type: archive
+        url: https://files.pythonhosted.org/packages/28/6c/640e3f5c734c296a7193079a86842a789edb7988dca39eab44579088a1d1/PyQt5-5.15.2.tar.gz
+        sha256: 372b08dc9321d1201e4690182697c5e7ffb2e0770e6b4a45519025134b12e4fc
+
+  - name: nlopt
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=Release
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DCMAKE_INSTALL_PREFIX=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:99]:
```suggestion
    buildsystem: cmake-ninja
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

File: com.ultimaker.cura.yml [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+
+
+  # Based on https://github.com/Ultimaker/cura-build-environment/blob/4.12/projects/base_requirements.txt
+  - python-modules/base-requirements.json
+
+  - name: protobuf
+    buildsystem: autotools
+    config-opts:
+      - --with-zlib=false
+      - --prefix=/app
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:66]:
```suggestion
```

File: com.ultimaker.cura.yml [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+      url: https://github.com/xianyi/OpenBLAS/archive/v0.3.13.tar.gz
+      sha256: 79197543b17cc314b7e43f7a33148c308b0807cd6381ee77f77e15acf3e6459e
+
+  - name: geos
+    buildsystem: cmake
+    config-opts:
+      - -DCMAKE_INSTALL_PREFIX=/app
+      - -DCMAKE_PREFIX_PATH=/app
+      - -DGEOS_ENABLE_TESTS=OFF
+      - -DCMAKE_BUILD_TYPE=Release
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:52]:
```suggestion
    buildsystem: cmake-ninja
    config-opts:
      - -DGEOS_ENABLE_TESTS=OFF
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by hfiguiere** [comment]
bot, build com.ultimaker.cura

File: python-modules/base-requirements.json [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+                    "type": "file",
+                    "only-arches": ["x86_64"],
+                    "url": "https://files.pythonhosted.org/packages/96/07/4d23f8e34e56d8eeb2c37cd5924928a01c3dd756a1d99e470181bc57551e/cryptography-3.4.8-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl",
+                    "sha256": "1eb7bb0df6f6f583dd8e054689def236255161ebbcf62b226454ab9ec663746b"
+                },
+                {
+                    "type": "file",
+                    "only-arches": ["aarch64"],
+                    "url": "https://files.pythonhosted.org/packages/f5/2e/a61d992645593be442b2666252f5d2fceb0777dfd2675a75d8914a353cd3/cryptography-3.4.8-cp36-abi3-manylinux_2_17_aarch64.manylinux2014_aarch64.whl",
+                    "sha265": "34dae04a0dce5730d8eb7894eab617d8a70d0c97da76b905de9efb7128ad7085"
```
Comment (hfiguiere) [inline python-modules/base-requirements.json:277]:
```suggestion
                    "sha256": "34dae04a0dce5730d8eb7894eab617d8a70d0c97da76b905de9efb7128ad7085"
```

**Comment by hfiguiere** [comment]
build fails because the appdata fail to validate:
```
builddir/files/share/appdata/com.ultimaker.cura.appdata.xml: FAILED:
• tag-missing           : <content_rating> required [use https://odrs.gnome.org/oars]
• url-not-found         : <screenshot> failed to download (HTTP 404: Not Found) [https://raw.githubusercontent.com/Ultimaker/Cura/master/screenshot.png]
• tag-missing           : <release> required
Validation of files failed
```

File: python-modules/base-requirements.json (hunk 21 above)
Comment (hfiguiere) [reply python-modules/base-requirements.json:277]: aarch64 build failed because of this

File: com.ultimaker.cura.yml (hunk 14 above)
Comment (Tim453) [reply com.ultimaker.cura.yml:149]: Thanks for your review! Why is this cleanup necessary? I thought it is enough to use the cleanup of the top module, which already removes the /include folder.

File: com.ultimaker.cura.yml (hunk 17 above)
Comment (hfiguiere) [reply com.ultimaker.cura.yml:105]: ah yeah right. I tend to add the to the modules, but the top-level should be good enough.

**Comment by nedrichards** [comment]
bot, build com.ultimaker.cura

you can use this command yourself as well when making changes

**Comment by hfiguiere** [comment]
Still needs appdata patches to not fail though

**Comment by hfiguiere** [comment]
I may have been excessive in what should be removed this time. :-/

File: com.ultimaker.cura.yml [hunk 22]
```diff
@@ -0,0 +1,285 @@
...
+        url: https://github.com/ultimaker/libSavitar.git
+        commit: 68a8868e0b30883b461c0dde77f6e85ce0a4d2ef
+        tag: 4.13.0
+      - type: patch
+        path: patch/libsavitar-python-install.patch
+
+  - name: libCharon
+    buildsystem: cmake-ninja
+    config-opts:
+      - -DPYTHON_EXECUTABLE=/app/bin/python3
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:218]: the option is actually listed as "not used" by CMake.

```suggestion
```

File: com.ultimaker.cura.yml [hunk 23]
```diff
@@ -0,0 +1,285 @@
...
+        url: https://github.com/ultimaker/libArcus.git
+        commit: 2066a9ddccece8f970902bb5e33147d674627edd
+        tag: 4.13.0 
+      - type: patch
+        path: patch/libarcus-python-binindg-install.patch
+
+  - name: pynest2d
+    buildsystem: cmake-ninja
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:190]:
```suggestion
      - -DCMAKE_PREFIX_PATH=/app
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

This actually fixes the build.

File: com.ultimaker.cura.yml [hunk 24]
```diff
@@ -0,0 +1,285 @@
...
+      - -DCMAKE_CXX_STANDARD=17
+      - -DBUILD_STATIC=ON
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:192]:
```suggestion
```

CMake says this is unused.

File: com.ultimaker.cura.yml [hunk 25]
```diff
@@ -0,0 +1,285 @@
...
+      - type: git
+        url: https://github.com/Ultimaker/libCharon
+        commit: 8e255aaad10b09c9cfc6e795f33968fd642e26c9
+        tag: 4.13.0
+
+  - name: cura-engine
+    buildsystem: cmake-ninja
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+      - -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:229]:
```suggestion
```

CMake say this is unused

**Review by hfiguiere — CHANGES_REQUESTED** [review]
I tested this locally

**Comment by hfiguiere** [comment]
My locall build works but I had to disable Wayland. It seems that there are bugs with Wayland, works fine with X11.

File: com.ultimaker.cura.yml [hunk 26]
```diff
@@ -0,0 +1,285 @@
...
+      - type: git
+        url: https://github.com/Ultimaker/libnest2d.git
+        commit: 031e4e61218edd092fbc54fbc6df145287b628ba
+      - type: file
+        url: https://raw.githubusercontent.com/Ultimaker/cura-build-environment/4.12/projects/libnest2d_find_clipper.cmake
+        sha256: 558d26c1071a4cb31c09e618553a3f98de7b782d779989db0582ab5c854a6d5f
+        dest-filename: FindClipper.cmake
+      - type: shell
+        commands:
+          - mv FindClipper.cmake cmake_modules/
```
Comment (hfiguiere) [inline com.ultimaker.cura.yml:140]:
```suggestion
        dest: cmake_modules/
```

This is much simpler.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by Tim453** [comment]
bot, build com.ultimaker.cura

**Comment by barthalion** [comment]
Hey @Ghostkeeper, I understand your stance of supporting only a single package for your app, but would you be interested in having write access to the Flathub repo? Just in case you would like to merge some PRs or close issues.

**Comment by Ghostkeeper** [comment]
You could add me as having write access if you like. But indeed I can't promise that I'll do anything with it.

If you/someone asks for help I'd be happy to lend my expertise though, and if it's easier to make a PR or close the issue then I'd use those permissions to do that.

So in short, I have no interest in spending time to maintain this. But I have good intentions and willingness to help with small things ;)

**Comment by barthalion** [comment]
/merge @Ghostkeeper

That's absolutely fine. Cheers :)

**Pull request closed without merging**